src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,387 @@
+        }
+
+        // the way the encoding is implemented in PowerShell 5 and earlier
+        // if the user sets the default encoding to WindowsLegacy, we will
+        // be able to encode for that
+        internal static Dictionary<String, Encoding> legacyEncodingMap =
+            new Dictionary<string, Encoding>(StringComparer.OrdinalIgnoreCase)
+            {
+                { "microsoft.powershell.commands.addcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportclixmlcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.exportcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportpssessioncommand", Encoding.UTF8 }, // with BOM
+                { "microsoft.powershell.commands.formathex", Encoding.ASCII },
+                { "microsoft.powershell.commands.newmodulemanifestcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.getcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.importcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.outfilecommand", Encoding.Unicode }, // This includes redirection
+                { "microsoft.powershell.commands.setcontentcommand", Encoding.ASCII },
+                // Providers are handled here
+                { "microsoft.powershell.commands.filesystemprovider", Encoding.ASCII },
+
+            };
+
+        internal static Encoding GetWindowsLegacyEncoding(string name)
+        {
+            if ( legacyEncodingMap.ContainsKey(name))
+            {
+                return legacyEncodingMap[name];
+            }
+            else 
+            {
+                return Encoding.Default;
+            }
+        }
+
+        /// <summary>
+        /// Retrieve the PSDefaultFileEncoding preference value if set
+        /// <summary>
+        public static FileEncoding GetEncodingPreference(SessionState sessionState)
+        {
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
+                object tmp = sessionState.PSVariable.GetValue("PSDefaultFileEncoding");
+                LanguagePrimitives.TryConvertTo<FileEncoding>(tmp, out encodingPreference);
+            }
+            catch
+            {
+                ;
+            }
+            return encodingPreference;
+        }
+
+        /// <summary>
+        /// Retrieve the encoding in a provider context
+        /// </summary>
+        public static Encoding GetProviderEncoding(CmdletProvider provider, FileEncoding encoding)
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = GetEncodingPreference(provider.SessionState);
+            if ( encoding == FileEncoding.Unknown && encodingPreference != FileEncoding.Unknown )
```

Comment:
yes - because of the provider cmdlets which leave the encoding to the underlying provider as a dynamic parameter it needs to be handled via a different path than checking against the name of the cmdlet.